addons/analytic/models/analytic.py

```diff
@@ -19,26 +19,28 @@ class account_analytic_account(models.Model):
 
     @api.multi
     def _compute_debit_credit_balance(self):
-        analytic_line_obj = self.env['account.analytic.line']
-        domain = [('account_id', 'in', self.mapped('id'))]
+        domain = [('account_id', 'in', self.ids),
+                  ('company_id', '=', self.env.user.company_id.id)]
         if self._context.get('from_date', False):
-            domain.append(('date', '>=', self._context['from_date']))
+            domain += [('date', '>=', self._context['from_date'])]
         if self._context.get('to_date', False):
-            domain.append(('date', '<=', self._context['to_date']))
-
-        account_amounts = analytic_line_obj.search_read(domain, ['account_id', 'amount'])
-        account_ids = set([line['account_id'][0] for line in account_amounts])
-        data_debit = {account_id: 0.0 for account_id in account_ids}
-        data_credit = {account_id: 0.0 for account_id in account_ids}
-        for account_amount in account_amounts:
-            if account_amount['amount'] < 0.0:
-                data_debit[account_amount['account_id'][0]] += account_amount['amount']
-            else:
-                data_credit[account_amount['account_id'][0]] += account_amount['amount']
+            domain += [('date', '<=', self._context['to_date'])]
+        # compute debits
+        debit_domain = domain + [('amount', '<', 0.0)]
+        debits = self.env['account.analytic.line'].read_group(
+            debit_domain, ['account_id', 'amount'], ['account_id'])
+        debits = dict([(amount['account_id'][0], amount['amount'])
```

Comment:
If you use a dictionary comprehension like `debits = {amount['account_id'][0]: amount['amount'] for amount in debits}` you might same some more time (admittedly, not much, but it also helps with readability):

```
$ ipython
In [1]: import random

In [2]: l = range(1000)

In [3]: random.shuffle(l)

In [4]: %timeit d = dict((x, x+3) for x in l)
1000 loops, best of 3: 250 µs per loop

In [5]: %timeit d = {x: x+3 for x in l}
10000 loops, best of 3: 151 µs per loop

In [6]: l = range(100000)

In [7]: random.shuffle(l)

In [8]: %timeit d = {x: x+3 for x in l}
10 loops, best of 3: 64.3 ms per loop

In [9]: %timeit d = dict((x, x+3) for x in l)
10 loops, best of 3: 76.5 ms per loop
```